File: 54-Spiral-Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new ArrayList<>();
        int top = 0, bottom = matrix.length - 1;
        int left = 0, right = matrix[0].length - 1;

        while (top <= bottom && left <= right) {
            // Move Right
            for (int i = left; i <= right; i++)
                result.add(matrix[top][i]);
            top++;

            // Move Down
            for (int i = top; i <= bottom; i++)
                result.add(matrix[i][right]);
            right--;

            // Move Left (only if rows remain)
            if (top <= bottom) {
                for (int i = right; i >= left; i--)
                    result.add(matrix[bottom][i]);
                bottom--;
            }

            // Move Up (only if columns remain)
            if (left <= right) {
                for (int i = bottom; i >= top; i--)
                    result.add(matrix[i][left]);
                left++;
            }
        }
        return result;
    }

}

File: https:/github.com/rajesh-kayal-dev/Leetcode.git/54-SpiralMatrix/54-SpiralMatrix.js
// Last updated: 7/25/2026, 10:59:38 AM
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let ans =[];

    let top = 0;
    let bottom = matrix.length - 1;

    let left = 0;
    let right = matrix[0].length - 1;

    while(top <= bottom && left <= right){
    //top row
    for(let i = left; i <= right; i++){
        ans.push(matrix[top][i]);
    }
    top++;

    //right column
    for(let i = top; i <= bottom; i++){
        ans.push(matrix[i][right]);
    }
    right--;

    //bottom row
    if(top <= bottom){
    for(let i = right; i >= left; i--){
        ans.push(matrix[bottom][i]);
    }
    bottom--;
    }

    //left column
    if(left <= right){
    for(let i = bottom; i >= top; i--){
        ans.push(matrix[i][left]);
    }
    left++;
    }
    }

    return ans;
    
};